packages/contract/src/contracts/contract.ts

```diff
@@ -10,19 +12,23 @@ import { FunctionInvocationScope } from './functions/invocation-scope';
 import { MultiCallInvocationScope } from './functions/multicall-scope';
 
 export default class Contract implements AbstractContract {
-  id!: string;
+  id!: AbstractAddress;
   provider!: Provider | null;
   interface!: Interface;
   wallet!: Wallet | null;
   functions: InvokeFunctions = {};
 
   constructor(
-    id: string,
+    id: string | AbstractAddress,
     abi: JsonAbi | Interface,
     walletOrProvider: Wallet | Provider | null = null
   ) {
     this.interface = abi instanceof Interface ? abi : new Interface(abi);
-    this.id = id;
+    if (id instanceof AbstractAddress) {
+      this.id = id;
+    } else {
+      this.id = isBech32(id) ? new Address(id) : fromB256(id);
```

Comment:
I think this can be something very common maybe we should abstract it inside the Address class.

```
Address.fromAny(string); // Something like this hehe
```